packages/components/src/navigation/components-composition-navigation.js

```diff
@@ -0,0 +1,141 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useState, useRef } from '@wordpress/element';
+import { Icon, chevronLeft, chevronRight } from '@wordpress/icons';
+
+/**
+ * Internal dependencies
+ */
+import Animate from '../animate';
+import {
+	BackButtonUI,
+	MenuItemTitleUI,
+	MenuItemUI,
+	MenuTitleUI,
+	MenuUI,
+	Root,
+} from './styles/navigation-styles';
+import Button from '../button';
+
+export default function ComponentsCompositionNavigation( {
+	initialActiveLevel,
+	children,
+} ) {
+	const [ activeLevel, setActiveLevel ] = useState( initialActiveLevel );
+	const [ slideOrigin, setSlideOrigin ] = useState( 'left' );
+
+	const isMounted = useRef( false );
+	useEffect( () => {
+		if ( ! isMounted.current ) {
+			isMounted.current = true;
+		}
+	}, [] );
+
+	const navigateTo = ( level ) => {
+		setActiveLevel( level );
+		setSlideOrigin( 'left' );
+	};
+
+	const navigateBack = ( level ) => {
+		setActiveLevel( level );
+		setSlideOrigin( 'right' );
```

Comment:
These methods are asynchronous, right?  Seems like a small risk they aren't updated prior to the level change.

Reply:
AFAIK set states are batched together before re-rendering, so level and slide origin should be both updated before the component replaces its content and perform the animation. 🤔

Reply:
Oh nice, you're right!  I thought this only applied to synthetic event handlers, but looks like calls will also be batched in `useEffect`. 👍